### PR TITLE
KEP-2238 probe-level grace period changes for beta/1.22

### PR DESCRIPTION
**Pull request by ehashman**
See https://github.com/kubernetes/enhancements/issues/2238#issuecomment-879178450

### CODE REVIEW

File: content/en/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes.md [hunk 1]
```diff
@@ -429,19 +429,28 @@ to resolve it.
...
-probe-level value.
+In 1.21 and beyond, when the feature flag `ProbeTerminationGracePeriod` is
+enabled, users can specify a probe-level `terminationGracePeriodSeconds` as
+part of the probe specification. When the feature flag is enabled, and both a
+pod- and probe-level `terminationGracePeriodSeconds` are set, the kubelet will
+use the probe-level value.
+
+{{< note >}}
+As of Kubernetes 1.22, the `ProbeTerminationGracePeriod` feature flag is only
+available on the API Server. The Kubelet will always honour the probe-level
```
Comment (sftim) [inline content/en/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes.md:448]:
```suggestion
available on the API Server. The kubelet always honors the probe-level
```

File: content/en/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes.md [hunk 2]
```diff
@@ -429,19 +429,28 @@ to resolve it.
...
+`terminationGracePeriodSeconds` field if it is present on a pod. To disable the
+feature on existing pods, they must be deleted and recreated. So long as the
```
Comment (sftim) [inline content/en/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes.md:450]:
```suggestion
`terminationGracePeriodSeconds` field if it is present on a pod. If you have
existing Pods where the `terminationGracePeriodSeconds` field is set and
you want to switch to per-probe termination grace periods, you must delete
those existing Pods.

When you (or the control plane, or some other component) create replacement
Pods, and the feature flag `ProbeTerminationGracePeriod` is enabled, then the
API server ignores the Pod-level `terminationGracePeriodSeconds` field, even if
a Pod or pod template specifies it.

So long as the `ProbeTerminationGracePeriod`
```
?

File: content/en/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes.md [hunk 3]
```diff
@@ -429,19 +429,28 @@ to resolve it.
...
+feature flag is off, the API Server will ignore the field on subsequent
```
Comment (sftim) [inline content/en/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes.md:451]: Did you mean:
```suggestion
feature flag is enabled, the API Server will ignore the field on subsequent
```
?

Comment (ehashman) [reply content/en/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes.md:451]: Should be disabled but gotcha

File: content/en/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes.md (hunk 1 above)
Comment (ehashman) [reply content/en/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes.md:448]: The :canada: in me is saddened :)

File: content/en/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes.md (hunk 2 above)
Comment (ehashman) [reply content/en/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes.md:450]: Taking this and rewording slightly.

File: content/en/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes.md [hunk 4]
```diff
@@ -429,19 +429,34 @@ to resolve it.
...
 coupling was unintended and may have resulted in failed containers taking an
 unusually long time to restart when a pod-level `terminationGracePeriodSeconds`
 was set.
 
-In 1.21, when the feature flag `ProbeTerminationGracePeriod` is enabled, users
-can specify a probe-level `terminationGracePeriodSeconds` as part of the probe
-specification. When the feature flag is enabled, and both a pod- and
-probe-level `terminationGracePeriodSeconds` are set, the kubelet will use the
-probe-level value.
+In 1.21 and beyond, when the feature flag `ProbeTerminationGracePeriod` is
```
Comment (tengqm) [inline content/en/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes.md:440]: Let's always call them 'feature gate'.

File: content/en/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes.md [hunk 5]
```diff
@@ -429,19 +429,34 @@ to resolve it.
...
+enabled, users can specify a probe-level `terminationGracePeriodSeconds` as
+part of the probe specification. When the feature flag is enabled, and both a
```
Comment (tengqm) [inline content/en/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes.md:442]: here too.

File: content/en/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes.md [hunk 6]
```diff
@@ -429,19 +429,34 @@ to resolve it.
...
+As of Kubernetes 1.22, the `ProbeTerminationGracePeriod` feature flag is only
+available on the API Server. The kubelet always honors the probe-level
+`terminationGracePeriodSeconds` field if it is present on a Pod.
+
+If you have existing Pods where the `terminationGracePeriodSeconds` field is set and
+you no longer wish to use per-probe termination grace periods, you must delete
+those existing Pods.
+
+When you (or the control plane, or some other component) create replacement
+Pods, and the feature flag `ProbeTerminationGracePeriod` is disabled, then the
```
Comment (tengqm) [inline content/en/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes.md:456]: here too

**Comment by sftim** [comment]
I haven't looked at how well this matches the code etc, but it's looking very good in terms of explaining things to a reader. Nice work.

**Comment by PI-Victor** [comment]
anything else to add here or can we merge this?

**Pull request merged**